File: leetcode/560.和为 K 的子数组/560.和为 K 的子数组.go
package main

func subarraySum(nums []int, k int) int {
	preSum := map[int]int{}
	preSum[0] = 1
	ans, sumI := 0, 0
	for _, num := range nums {
		sumI += num
		sumJ := sumI - k
		if _, ok := preSum[sumJ]; ok {
			ans += preSum[sumJ]
		}
		preSum[sumI] += 1
	}
	return ans
}

func main() {
	nums := []int{1, 2, 3}
	k := 3
	result := subarraySum(nums, k)
	println(result)
}

File: leetcode/560.和为 K 的子数组/560.和为 K 的子数组.py
from typing import List


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        """
        pre_sum_map 记录前缀和出现次数，用于统计, base case {0:1}
        sum_j = sum_i - k 用两数和思路，找到适合的前缀和 sum_j
        从前缀和map找到次数
        """
        pre_sum_map = {0: 1}
        ans = sum_i = 0
        for num in nums:
            sum_i += num
            sum_j = sum_i - k
            # 存在就 ans+1
            if sum_j in pre_sum_map:
                ans += pre_sum_map[sum_j]
            # 更新前缀和map
            pre_sum_map[sum_i] = pre_sum_map.get(sum_i, 0) + 1
        return ans


if __name__ == '__main__':
    nums = [1, 1, 1]
    k = 2
    result = Solution().subarraySum(nums, k)
    print(result)
